src/celeritas/global/KernelContextException.cc

```diff
@@ -0,0 +1,164 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/global/KernelContextException.cc
+//---------------------------------------------------------------------------//
+#include "KernelContextException.hh"
+
+#include "corecel/cont/ArrayIO.hh"
+#include "corecel/io/JsonPimpl.hh"
+#if CELERITAS_USE_JSON
+#    include "corecel/cont/Array.json.hh"
+#    include "corecel/math/QuantityIO.json.hh"
+#endif
+
+#include "CoreTrackView.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+#if CELERITAS_USE_JSON
+template<class V, class S>
+void insert_if_valid(char const* key,
+                     OpaqueId<V, S> const& val,
+                     nlohmann::json* obj)
+{
+    if (val)
+    {
+        (*obj)[key] = val.unchecked_get();
+    }
+}
+#endif
+template<class V, class S>
+std::ostream& operator<<(std::ostream& os, OpaqueId<V, S> const& v)
+{
+    if (v)
+    {
+        os << v.unchecked_get();
+    }
+    else
+    {
+        os << '?';
+    }
+    return os;
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with track data and kernel label.
+ */
+KernelContextException::KernelContextException(CoreHostRef const& data,
+                                               ThreadId thread,
+                                               std::string&& label)
+    : thread_(thread), label_(std::move(label))
+{
+    try
+    {
+        if (!thread)
+        {
+            // Make sure the thread is valid before trying to construct
+            // detailed debug informatino
+            throw std::exception();
+        }
+        CoreTrackView core(data.params, data.states, thread);
+        this->initialize(core);
+    }
+    catch (...)
+    {
+        // Ignore all exceptions while trying to process diagnostic information
```

Comment:
Is that because they really don't matter or because it is challenging to capture any information about them?

Reply:
They might matter, but this constructor is almost always happening inside a `catch` block of some other failure so it's likely that some parts of the state are indeterminate... so I think it's better to just ignore them and report the original failure. I'm not sure what much else we could do.

Reply:
> but this constructor is almost always happening inside a ..... what much else we could do.

I guess, at least at this information/assumption in the code comment.